ql/src/test/results/clientpositive/llap/swo_event_merge.q.out

```diff
@@ -0,0 +1,291 @@
+	and d.d_year=2000
+union
+select   s.ss_item_sk*d_date_sk
+ from
+     x1_store_sales s
+     ,x1_date_dim d
+ where  
+	1=1
+	and s.ss_sold_date_sk = d.d_date_sk
+	and d.d_year=2001
+	group by s.ss_item_sk*d_date_sk
+PREHOOK: type: QUERY
+PREHOOK: Input: default@x1_date_dim
+PREHOOK: Input: default@x1_store_sales
+PREHOOK: Input: default@x1_store_sales@ss_sold_date_sk=1
+PREHOOK: Input: default@x1_store_sales@ss_sold_date_sk=2
+#### A masked pattern was here ####
+POSTHOOK: query: explain 
+select   count(*) cnt
+ from
+     x1_store_sales s
+     ,x1_date_dim d
+ where  
+	1=1
+	and s.ss_sold_date_sk = d.d_date_sk
+	and d.d_year=2000
+union
+select   s.ss_item_sk*d_date_sk
+ from
+     x1_store_sales s
+     ,x1_date_dim d
+ where  
+	1=1
+	and s.ss_sold_date_sk = d.d_date_sk
+	and d.d_year=2001
+	group by s.ss_item_sk*d_date_sk
+POSTHOOK: type: QUERY
+POSTHOOK: Input: default@x1_date_dim
+POSTHOOK: Input: default@x1_store_sales
+POSTHOOK: Input: default@x1_store_sales@ss_sold_date_sk=1
+POSTHOOK: Input: default@x1_store_sales@ss_sold_date_sk=2
+#### A masked pattern was here ####
+Plan optimized by CBO.
+
+Vertex dependency in root stage
+Reducer 2 <- Map 1 (SIMPLE_EDGE), Map 8 (SIMPLE_EDGE)
+Reducer 3 <- Reducer 2 (CUSTOM_SIMPLE_EDGE), Union 4 (CONTAINS)
+Reducer 5 <- Union 4 (SIMPLE_EDGE)
+Reducer 6 <- Map 1 (SIMPLE_EDGE), Map 8 (SIMPLE_EDGE)
+Reducer 7 <- Reducer 6 (SIMPLE_EDGE), Union 4 (CONTAINS)
+
+Stage-0
+  Fetch Operator
+    limit:-1
+    Stage-1
+      Reducer 5 vectorized, llap
+      File Output Operator [FS_89]
+        Group By Operator [GBY_88] (rows=1 width=8)
+          Output:["_col0"],keys:KEY._col0
+        <-Union 4 [SIMPLE_EDGE]
+          <-Reducer 3 [CONTAINS] vectorized, llap
+            Reduce Output Operator [RS_87]
+              PartitionCols:_col0
+              Group By Operator [GBY_86] (rows=1 width=8)
+                Output:["_col0"],keys:_col0
+                Group By Operator [GBY_85] (rows=1 width=8)
+                  Output:["_col0"],aggregations:["count(VALUE._col0)"]
+                <-Reducer 2 [CUSTOM_SIMPLE_EDGE] llap
+                  PARTITION_ONLY_SHUFFLE [RS_11]
+                    Group By Operator [GBY_10] (rows=1 width=8)
+                      Output:["_col0"],aggregations:["count()"]
+                      Merge Join Operator [MERGEJOIN_51] (rows=1728398 width=8)
+                        Conds:RS_71._col0=RS_77._col0(Inner)
+                      <-Map 1 [SIMPLE_EDGE] vectorized, llap
+                        SHUFFLE [RS_71]
+                          PartitionCols:_col0
+                          Select Operator [SEL_69] (rows=123457 width=4)
+                            Output:["_col0"]
+                            Filter Operator [FIL_68]
+                              predicate:ss_sold_date_sk is not null
+                              TableScan [TS_0] (rows=123457 width=14)
+                                default@x1_store_sales,s,Tbl:COMPLETE,Col:COMPLETE,Output:["ss_item_sk"]
```

Comment:
Could merging these TS operators be effectively worse?

For instance, in this specific mock query, no partition will be pruned for the *x1_store_sales* table, while before partition pruning was kicking in. Thus, in this case, you are scanning the same data whether you have one or two TS operators (two partitions), however after merging the TS, the size of the data you are shuffling for the join doubles (data in both partitions twice)? Is that analysis correct?

Off the top of my head, this could be beneficial if i) both TS only select a small subset of the partitions in the table, or ii) overlapping in the partition list for those two different TS is greater than a certain threshold. Should we work in that direction, i.e., introduce some config parameters for this?

@rbalamohan , what is your take? It would be helpful to have a second opinion.

Reply:
I think right now we don't have a sanity check filter right before the TS to limit reduce shuffled data size to the previous amount; but reducing the number of scans is beneficial - IIRC in the q23 query the scanned partitions were the same; so the benefit was real.

If we make available these things the same way as we have the SJ filters - then we could for sure avoid shuffling more data.

Note: I think it would enable some further opportunities if we would change the SJ data transmission method from RS to EVENTOP - that way we shouldn't have to worry about parallel edges anymore...

Reply:
@jcamachor should I add a config knob to control event operator merge?

Reply:
>> however after merging the TS, the size of the data you are shuffling for the join doubles


Yes, this is possible with merging. +1 on adding an option for enabling it. 

There is additional option to reduce number of group by operator evaluations (e.g [Q65](https://github.com/apache/hive/blob/master/ql/src/test/queries/clientpositive/perf/query65.q#L18) computes group by operator twice) which can help in reducing runtime.